### PR TITLE
Vault metrics

### PR DESCRIPTION
**Pull request by robobario**
### Type of change

- Enhancement / new feature

### Description

We want to start making it visible how many API calls are being made to Vault and how many operations are resulting in keys not being found or other exceptions.

Example of the new metrics obtained via the prometheus endpoint:
```
kroxylicious_kms_vault_request_attempt_total{operation="resolve_alias",} 1.0
kroxylicious_kms_vault_request_attempt_total{operation="decrypt_edek",} 1.0
kroxylicious_kms_vault_request_attempt_total{operation="create_datakey",} 1.0
kroxylicious_kms_vault_request_outcome_total{failure="",operation="create_datakey",outcome="success",} 1.0
kroxylicious_kms_vault_request_outcome_total{failure="not_found",operation="decrypt_edek",outcome="failure",} 0.0
kroxylicious_kms_vault_request_outcome_total{failure="",operation="decrypt_edek",outcome="success",} 1.0
kroxylicious_kms_vault_request_outcome_total{failure="exception",operation="create_datakey",outcome="failure",} 0.0
kroxylicious_kms_vault_request_outcome_total{failure="exception",operation="resolve_alias",outcome="failure",} 0.0
kroxylicious_kms_vault_request_outcome_total{failure="not_found",operation="resolve_alias",outcome="failure",} 0.0
kroxylicious_kms_vault_request_outcome_total{failure="not_found",operation="create_datakey",outcome="failure",} 0.0
kroxylicious_kms_vault_request_outcome_total{failure="exception",operation="decrypt_edek",outcome="failure",} 0.0
kroxylicious_kms_vault_request_outcome_total{failure="",operation="resolve_alias",outcome="success",} 1.0
```
The idea is that the sum of the outcomes series for an operation should equal the attempt count for that operation.

Note that this is a limited set of metrics that are registered on startup so that we can start emitting 0.0 values, we do not tag by HTTP status code etc. This helps systems like prometheus alert on changes in each series, there can be issues if you emit no data until the very first occurrence of your error.

We could add HTTP status breakdown in a subsequent PR, or rely on logging to debug failures.

### Checklist

- [ ] Write tests
- [ ] Make sure all tests pass
- [ ] Review performance test results. Ensure that any degradations to performance numbers are understood and justified.
- [ ] Make sure all Sonar-Lint warnings are addressed or are justifiably ignored.
- [ ] Update documentation
- [ ] Reference relevant issue(s) and close them after merging
- [ ] For user facing changes, update CHANGELOG.md (remember to include changes affecting the API of the test artefacts too).

### CODE REVIEW

**Comment by robobario** [comment]
I ended up making these first metrics at a logical level above the level of HTTP request and response counts so that there were fewer counters that could be emitted immediately on startup.

Maybe it would be better to implement this in the Filter? We could wrap these measurements around the KMS regardless of implementation.

**Comment by SamBarker** [comment]
Feels like we would want metrics at the filter level for counting requests/found/missing etc thats the "business" metrics we care about, HTTP codes etc are implementation details. They should in theory only be interesting in diagnosing issues with how we translate errors or the actual KMS systems.

File: kroxylicious-kms-provider-hashicorp-vault/src/main/java/io/kroxylicious/kms/provider/hashicorp/vault/VaultKms.java [hunk 1]
```diff
@@ -83,12 +83,11 @@ private HttpClient createClient() {
     @NonNull
     @Override
     public CompletionStage<DekPair<VaultEdek>> generateDekPair(@NonNull String kekRef) {
-
         var request = createVaultRequest()
                 .uri(vaultUrl.resolve("v1/transit/datakey/plaintext/%s".formatted(encode(kekRef, UTF_8))))
                 .POST(HttpRequest.BodyPublishers.noBody())
                 .build();
-
+        Metrics.createDataKey().attempt().increment();
```
Comment (SamBarker) [inline kroxylicious-kms-provider-hashicorp-vault/src/main/java/io/kroxylicious/kms/provider/hashicorp/vault/VaultKms.java:90]: the method being counted is `generate` so why is the metric `create` ?

File: kroxylicious-kms-provider-hashicorp-vault/src/main/java/io/kroxylicious/kms/provider/hashicorp/vault/VaultKms.java [hunk 2]
```diff
@@ -97,6 +96,8 @@ public CompletionStage<DekPair<VaultEdek>> generateDekPair(@NonNull String kekRe
                 .thenApply(data -> {
                     var secretKey = new SecretKeySpec(Base64.getDecoder().decode(data.plaintext()), AES_KEY_ALGO);
                     return new DekPair<>(new VaultEdek(kekRef, data.ciphertext().getBytes(UTF_8)), secretKey);
+                }).whenComplete((vaultEdekDekPair, throwable) -> {
+                    recordMetrics(throwable, Metrics.createDataKey());
```
Comment (SamBarker) [inline kroxylicious-kms-provider-hashicorp-vault/src/main/java/io/kroxylicious/kms/provider/hashicorp/vault/VaultKms.java:100]: nit: swap the argument order as I think it reads better.

**Comment by robobario** [comment]
> Feels like we would want metrics at the filter level for counting requests/found/missing etc thats the "business" metrics we care about, HTTP codes etc are implementation details. They should in theory only be interesting in diagnosing issues with how we translate errors or the actual KMS systems.

I think I'll rework this to be at the filter level as this is going to be more interesting than the KMS implementation details at first.

**Comment by robobario** [comment]
Superceded by #861

**Pull request closed without merging**